src/compiler/checker.ts

```diff
@@ -14908,6 +14908,12 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (type.flags & TypeFlags.Object) {
             const resolved = resolveStructuredTypeMembers(type as ObjectType);
             const symbol = resolved.members.get(name);
+            if (symbol && !includeTypeOnlyMembers && type.symbol?.flags & SymbolFlags.ValueModule && getSymbolLinks(type.symbol).typeOnlyExportStarMap?.has(name)) {

```

Comment:
I don't understand the check for `!includeTypeOnlyMembers`. If this is from a `import * as intermediate from './x.js'`, the type doesn't indicate that type-only members should be skipped. Where does this come from?

Actually, the overall intent is to skip values that show up in the type-only exports map, right? How does this code distinguish a value (which shouldn't be included) from a type (which should (I think?))? Maybe that waits until the check just below the new one.

Reply:
This function gets _properties_ which need to be values by definition. The question is whether the value has to exist in the emitted JavaScript or not. The only callers that pass `includeTypeOnlyMembers: true` are type queries and import type nodes:

```ts
type T = typeof x.y.z; // 'y' may be a namespace made by `export type * from "y"`
type U = typeof import("x").y.z;
```

These are constructs that query values, but don’t rely on those values actually existing, so they can safely return value symbols that are only accessible via type-only imports/exports.

Reply:
Thanks, that example helped a lot. So the namespace 'y' can be, almost, a synthetic property that should only be visible from type space?

Reply:
I didn’t show where x/y/z were declared, but the way I intended the example was that `y` itself is real (it’s the module namespace object of a file that does `export type *` but maybe has no real exports—so it would be an empty object), while `z` is the thing that acts like a “synthetic property” on `y`. But I think you understood the idea.